internal/nodes/avi_vrf_translator.go

```diff
@@ -31,42 +29,128 @@ import (
+			nodeStaticRoute.Count = len(nodeRoutes)
+			aviVrfNode.NodeStaticRoutes[nodeName] = nodeStaticRoute
+			aviVrfNode.Nodes = append(aviVrfNode.Nodes, nodeName)
+		}
+	} else if !deleteFlag {
+		// update case
+		// Assumption: updated routes (values) for given node will not overlap with other nodes.
+		// So only updating existing routes of that node.
+		startIndex := nodeStaticRouteDetails.StartIndex
+		lenNewNodeRoutes := len(nodeRoutes)
+		diff := lenNewNodeRoutes - nodeStaticRouteDetails.Count
+
+		var staticRouteCopy []*models.StaticRoute
+		staticRouteCopy = append(staticRouteCopy, aviVrfNode.StaticRoutes[:startIndex]...)
+		staticRouteCopy = append(staticRouteCopy, nodeRoutes...)
+
+		staticRouteCopy = append(staticRouteCopy, aviVrfNode.StaticRoutes[startIndex+nodeStaticRouteDetails.Count:]...)
+		aviVrfNode.StaticRoutes = staticRouteCopy
+
+		//if diff is 0, there is no change in number of routes previously exist and newly created.
+		if diff != 0 {
+			updateNodeStaticRoutes(aviVrfNode, deleteFlag, nodeName, lenNewNodeRoutes, diff)
 		}
+
+	} else {
+		//delete case
+		startIndex := nodeStaticRouteDetails.StartIndex
+		count := nodeStaticRouteDetails.Count
+		var staticRouteCopy []*models.StaticRoute
+		staticRouteCopy = append(staticRouteCopy, aviVrfNode.StaticRoutes[:startIndex]...)
+		staticRouteCopy = append(staticRouteCopy, aviVrfNode.StaticRoutes[startIndex+count:]...)
+		aviVrfNode.StaticRoutes = staticRouteCopy
+		countToSubstract := aviVrfNode.NodeStaticRoutes[nodeName].Count
+		updateNodeStaticRoutes(aviVrfNode, deleteFlag, nodeName, 0, -countToSubstract)
 	}
 	aviVrfNode.CalculateCheckSum()
-	o.AddModelNode(aviVrfNode)
 	utils.AviLog.Infof("key: %s, Added vrf node %s", key, vrfName)
 	utils.AviLog.Infof("key: %s, Number of static routes %v", key, len(aviVrfNode.StaticRoutes))
 	return nil
 }
 
+func updateNodeStaticRoutes(aviVrfNode *AviVrfNode, isDelete bool, nodeName string, lenNewNodeRoutes, diff int) {
+	//get index of nodename in node array
+	index := -1
+	for i := 0; i < len(aviVrfNode.Nodes); i++ {
+		if aviVrfNode.Nodes[i] == nodeName {
+			index = i
+			if !isDelete {
+				nodeNameToUpdate := aviVrfNode.Nodes[index]
+				nodeDetails := aviVrfNode.NodeStaticRoutes[nodeNameToUpdate]
+				nodeDetails.Count = lenNewNodeRoutes
+				aviVrfNode.NodeStaticRoutes[nodeNameToUpdate] = nodeDetails
+			}
+			break
+		}
+	}
+	if index != -1 {
+		//Change nodemap entries till index
+		for i := len(aviVrfNode.Nodes) - 1; i > index; i-- {
+			nodeNameToUpdate := aviVrfNode.Nodes[i]
+			nodeDetails := aviVrfNode.NodeStaticRoutes[nodeNameToUpdate]
+			nodeDetails.StartIndex = nodeDetails.StartIndex + diff
+			aviVrfNode.NodeStaticRoutes[nodeNameToUpdate] = nodeDetails
+		}
+		if isDelete {
+			// now remove nodename from Nodes list
+			updateNodeList := aviVrfNode.Nodes[:index]
+			if index+1 < len(aviVrfNode.Nodes) {
+				updateNodeList = append(updateNodeList, aviVrfNode.Nodes[index+1:]...)
```

Comment:
Do we have to take care of deleting the details from `NodeStaticRoutes` also?

Reply:
yes. taken care of deleting nodename entry from NodeStaticRoutes at line number 114.